shimmingtoolbox/shim/sequencer.py

```diff
@@ -23,857 +21,1365 @@
+            nib.save(nii_riro, os.path.join(self.path_output, 'fig_riro_fmap_component.nii.gz'))
+        return static, riro, mean_p, pressure_rms
+
+    def shim(self):
+        """
+        Performs realtime shimming using one of the supported optimizers and an external respiratory trace.
+
+        Returns:
+            (tuple): tuple containing:
+                * numpy.ndarray: Static coefficients of the coil profiles to shim (len(slices) x channels) e.g. [Hz]
+                * numpy.ndarray: Riro coefficients of the coil profiles to shim (len(slices) x channels)
+                  e.g. [Hz/unit_pressure]
+                * float: Mean pressure of the respiratory trace.
+                * float: Root mean squared of the pressure.
+                    This is provided to compare results between scans, multiply the riro coefficients
+                    by rms of the pressure to do so.
+        """
+
+        affine_fieldmap = self.nii_fieldmap.affine
+        static, riro, mean_p, pressure_rms = self.get_real_time_parameters()
+        # Static shim
+        self.select_optimizer(static, affine_fieldmap)
+        logger.info("Static optimization")
+        coef_static = self.optimize(self.nii_static_mask)
+
+        # RIRO optimization
+        # Use the currents to define a list of new coil bounds for the riro optimization
+        self.bounds = new_bounds_from_currents(coef_static, self.optimizer.merged_bounds)
+
+        if self.method == 'least_squares':
+            self.method = 'least_squares_rt'
+
+        self.select_optimizer(riro, affine_fieldmap, self.pmu)
+        logger.info("Realtime optimization")
+        coef_riro = self.optimize_riro(self.nii_riro_mask)
+
+        # Multiplying by the RMS of the pressure allows to make abstraction of the tightness of the bellow
+        # between scans. This allows to compare results between scans.
+        # coef_riro_rms = coef_riro * pressure_rms
+        # [unit_shim/unit_pressure] * rms_pressure, ex: [Hz/unit_pressure] * rms_pressure
+
+        return coef_static, coef_riro, mean_p, pressure_rms
+
+    def select_optimizer(self, unshimmed, affine, pmu: PmuResp = None):
+        """
+        Select and initialize the optimizer
+
+        Args:
+            unshimmed (numpy.ndarray): 3D B0 map
+            affine (numpy.ndarray): 4x4 array containing the affine transformation for the unshimmed array
+            pmu (PmuResp): PmuResp object containing the respiratory trace information. Required for method
+                           'least_squares_rt'.
+        Returns:
+            Optimizer : Initialized Optimizer object
+        """
+
+        # global supported_optimizers
+        if self.method in supported_optimizers:
+            if self.method == 'least_squares':
+                optimizer = supported_optimizers[self.method](self.coils, unshimmed, affine, self.opt_criteria,
+                                                              reg_factor=self.reg_factor)
```

Comment:
This will never get executed because if it's realtime, it will use the realtime optimizer:
```
        if self.method == 'least_squares':
            self.method = 'least_squares_rt'
```

We can remove it

Reply:
I'm not sure about that, because in the case of a RT shimming, we are first doing a static optimization and then an optimization with the Riro ? If it's the case then we have to make sure to do both optimization ? (I am not sure about that)